### DISCUSSION TITLE
Databases redux

### DISCUSSION THREAD

**Discussion by MikeSpreitzer** [kubestellar/kubestellar#596] [Ideas]
The biggest scale challenge in edge-mc is the reported state currently collected in the mailbox workspaces.  There are a few approaches we might take to coping with that.

One might be the sharding for scale that is in kcp-core.  I recently learned that there actually is such a thing.  I had misunderstood the proxy involved in kcp-core sharding, thinking that it is a barrier to scaling out.  I am told that I was wrong about that.  I do not know much more yet.

Another might be to resurrect Steve Kuznetsov's work on direct-to-CRDB storage in kcp-core.

Another might be to combine (a) switching from mailbox workspaces to wrapped objects and (b) get Steve's direct-to-CRDB storage working in plain Kubernetes.  That is a heavier lift, getting his work into upstream Kubernetes.  A possibly easier variant on this idea would be to resurrect Simon Seif's DB-based kube-ish apiserver and switch its storage from DynamoDB to the DB of our choice.  This might be feasible because this apiserver only needs to be good enough for three clients: the edge-mc syncer, placement translator, and mailboxwatch library.

Another approach would be to take those three clients and modify them to be direct DB clients rather than kube api clients.

**Comment by pdettori** [comment]
For the storage scaling, we may want to consider kine + Postgres as well, which would require minimal effort and is already supported in [kubeflex](https://github.com/kubestellar/kubeflex).

**Comment by MikeSpreitzer** [comment]
We could roadmap this.  Start with switching from mailbox workspaces to mailbox namespaces of wrapped objects, and later think about how to scale out the storage of the wrapped objects.  The down-side is that this skips past the ideas around scaling out storage of non-wrapped objects.  We could eventually look at scaling out storage of non-wrapped objects, but that would discard the implementation work based on wrapped objects and necessitate a discontinuity for clients.

**Comment by MikeSpreitzer** [comment]
For direct-to-DB storage, this raises the question of the mapping between the JSON/protobuf representation and the SQL representation.  The following approaches occur to me.

1. Use a fixed, normalized, indexed representation for the `TypeMeta` and `ObjectMeta` sections and use the DBMS's JSON datatype for the rest.  This would enable efficient SQL queries on the TypeMeta and ObjectMeta and doable but inefficient querying on the full objects.
2. Like 1 but extend the normalized and indexed part of the representation to cover more of the object.  This raises a couple of questions: (a) what about diversity of the resource's schema between apiservers and (b) who decides what part(s) of the object to normalize and index and how are those decisions conveyed?  Question (b) mitigates (a) --- the parts that remain in JSON do not necessarily need to agree on schema.  More importantly, the existing design for version unification mitigates (a).  But it is only mitigation --- different releases of Kubernetes have different schemas, in general, for the same group.version.resource.
